File: Depth_First_Search/1376.Time_Needed_to_Inform_All_Employees/Python/1376.Time_Needed_to_Inform_All_Employees.py
class Solution(object):
    def numOfMinutes(self, n, headID, manager, informTime):
        """
        :type n: int
        :type headID: int
        :type manager: List[int]
        :type informTime: List[int]
        :rtype: int
        """
        #============================================#
        # Recursion-based DFS & BFS traversal method #
        #============================================#

        ############
        #Initialize
        ##### Record dictionary/hashmap #####
        record_dict = {}

        ###########################################################################
        #Recursion-based DFS & BFS loop traversal with recorded dictionary/hashmap
        for n_idx in range(n):

            ##### Check if the current indexed-value existed or not #####
            if (manager[n_idx] not in record_dict):
                record_dict[manager[n_idx]] = [[n_idx, informTime[n_idx]]] #Keep updating/recording
            else:
                (record_dict[manager[n_idx]]).append([n_idx, informTime[n_idx]]) #Keep updating/recording

        return self.dfsbfsTraversal(record_dict, (-1), 0) #Recursion function call


    def dfsbfsTraversal(self, record_dict, record_manager, record_time):
        """
        :type record_dict: dict
        :type record_manager: int
        :type record_time: int
        :rtype: int
        """
        #==============================================#
        # Recursion-based DFS & BFS traversal function #
        #==============================================#
        if (record_manager not in record_dict): #Issue/Boundary-case handling
            return record_time

        ############
        #Initialize
        ##### Result maximun time #####
        res_max_time = record_time

        ####################
        #Whole process/flow
        for [sub_employee, sub_time] in record_dict[record_manager]:

            tmp_time = self.dfsbfsTraversal(record_dict, sub_employee, (record_time + sub_time)) #Recursion function call

            ##### Check if the current time matched conditions or not #####
            if (tmp_time > res_max_time):
                res_max_time = tmp_time #Keep updating/overwriting
            else:
                pass

        return res_max_time

File: Depth_First_Search/1376.Time_Needed_to_Inform_All_Employees/C++/1376.Time_Needed_to_Inform_All_Employees.cpp
class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int> & manager, vector<int> & informTime) 
    {
        //=======================================//
        // Input type:                           //
        //   - int (n employees)                 //
        //   - int (headID)                      //
        //   - vector<int> ref. (manager vector) //
        //   - vector<int> & (informTime vector) //
        // Return type:                          //
        //   - int (result maximun time)         //
        //=======================================//

        /*Initialize*/
        ///// Record dictionary, hashmap /////
        unordered_map<int, vector<pair<int, int>>> record_map;

        /*Recursion-based DFS loop traversal with recorded dictionary, hashmap*/
        for (int n_idx = 0; (n_idx < n); (n_idx++)) //Whole
        {
            (record_map[manager[n_idx]]).emplace_back(make_pair(n_idx, informTime[n_idx])); //Keep updating, recording

        } //Whole

        return dfsTraversal(record_map, (-1), 0); //Recursion function call
    }


    int dfsTraversal(unordered_map<int, vector<pair<int, int>>> & record_map, int record_manager, int record_time)
    {
        //==================================================================================//
        // Input type:                                                                      //
        //   - unordered_map<int, vector<pair<int, int>>> ref. (record dictionary, hashmap) //
        //   - int (record manager)                                                         //
        //   - int (record time)                                                            //
        // Return type:                                                                     //
        //   - int (result maximun time)                                                    //
        //==================================================================================//
        if (record_map.find(record_manager) == record_map.end()) { return record_time; } //Issue, Boundary-case handling

        /*Initialize*/
        ///// Temporal, Result maximun time /////
        int tmp_time, res_max_time = record_time;

        /*Whole process, flow*/
        for (pair<int, int> record_pair: record_map[record_manager]) //Whole
        {
            tmp_time = dfsTraversal(record_map, record_pair.first, (record_time + record_pair.second)); //Recursion function call

            res_max_time = (tmp_time > res_max_time) ? tmp_time : res_max_time; //Keep updating, overwriting

        } //Whole

        return res_max_time;
    }
};
